fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,635 @@
+     */
+    public static class SizeStatisticsResults {
+        private long keyCount;
+        private long keySize;
+        private long maxKeySize;
+        private long valueSize;
+        private long maxValueSize;
+        private long[] sizeBuckets;
+
+        private SizeStatisticsResults() {
+            this.keyCount = 0;
+            this.keySize = 0;
+            this.maxKeySize = 0;
+            this.valueSize = 0;
+            this.maxValueSize = 0;
+            this.sizeBuckets = new long[Integer.SIZE];
+        }
+
+        /**
+         * Update partial kv stats.
+         *
+         * @param kv latest key and value for partial stats update
+         */
+        private void updateStatistics(@Nonnull KeyValue kv) {
+            this.keyCount += 1;
+            this.keySize += kv.getKey().length;
+            this.maxKeySize = Math.max(this.maxKeySize, kv.getKey().length);
+            this.valueSize += kv.getValue().length;
+            this.maxValueSize = Math.max(this.maxValueSize, kv.getValue().length);
+            int totalSize = kv.getKey().length + kv.getValue().length;
+            if (totalSize > 0) {
+                this.sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
+            }
+        }
+
+        /**
+         * Sets partial kv stats from a protobuf serialization of partial stats results.
+         *
+         * @param partialResults partial kv stats from a protobuf serialization of partial stats results.
+         */
+        public void setFromProto(RecordCursorProto.SizeStatisticsPartialResults partialResults) {
+            this.keyCount = partialResults.getKeyCount();
+            this.keySize = partialResults.getKeySize();
+            this.maxKeySize = partialResults.getMaxKeySize();
+            this.valueSize = partialResults.getValueSize();
+            this.maxValueSize = partialResults.getMaxValueSize();
+            List<Long> bucketList = partialResults.getSizeBucketsList();
+            Arrays.setAll(sizeBuckets, i -> bucketList.get(i));
+        }
+
+        /**
+         * Return a copy of this instance.
+         *
+         * @return a copy of this instance
+         */
+        public SizeStatisticsResults copy() {
+            SizeStatisticsResults target = new SizeStatisticsResults();
+            target.setKeyCount(this.getKeyCount());
+            target.setKeySize(this.getKeySize());
+            target.setMaxKeySize(this.getMaxKeySize());
+            target.setValueSize(this.getValueSize());
+            target.setMaxValueSize(this.getMaxValueSize());
+            target.setSizeBuckets(this.getSizeBuckets());
+            return target;
+        }
+
+        /**
+         * Set the total number of keys in the requested key range.
+         *
+         * @param keyCount total number of keys in the requested key range.
+         */
+        public void setKeyCount(long keyCount) {
+            this.keyCount = keyCount;
```

Comment:
Are these modifiers used by things outside of the `copy` method? Given how one probably wants to mutate several of the statistics being collected here at once, I'm not sure we want to expose this as part of the public API.

Reply:
I am not super happy with this SizeStatisticsResult class. Ideally it would be something that would only get burped out by SizeStatisticsCollectorCursor.onNext(). And the only public methods would be the stats getters. However, because of the SizeStatisticsCollector usage there is a static method to get an initialized version of one of these i.e. SizeStatisticsCollectorCursor.getSizeStatisticsResultsInitialState(). So since you can get your hands on an initial version of one outside of running the cursor it seemed to make sense that there should be a way to set the base statistics. Perhaps you might want to do it for testing purposes? E.g. testing the higher level say P95 getters based on setting base stats? Anyway, it does seem a little funky but that was the argument I had with myself to expose them.

Reply:
So the resolution after communicating with you is to make the SizeStatisticsResults data structure read-only and to remove the old SizeStatisticsCollector from the public by dissolving it into  SizeStatisticsCollectorTest. Please see the latest branch and give me the green check mark : )